### ISSUE TITLE
Couchbase: Integration tests not launched with last JHipster dependency version

### ISSUE THREAD

**Issue by hdurix** [jhipster/generator-jhipster#7654]
##### **Overview of the issue**

Integration tests with Couchbase can not be launched: an error of a refused connection is thrown.

The stacktrace is:

```
2018-05-18 17:11:09.481  WARN 19481 --- [      cb-io-1-4] c.couchbase.client.core.RequestHandler   : Received Error during Reconfiguration.

com.couchbase.client.deps.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:11210
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717)
	at com.couchbase.client.deps.io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:225)
	at com.couchbase.client.deps.io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:291)
	at com.couchbase.client.deps.io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:634)
	at com.couchbase.client.deps.io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:581)
	at com.couchbase.client.deps.io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:498)
	at com.couchbase.client.deps.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:460)
	at com.couchbase.client.deps.io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:131)
	at com.couchbase.client.deps.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.ConnectException: Connection refused
	... 11 common frames omitted

```

##### **Reproduce the error**

- Generate a project with Couchbase as database from the master branch.
- Run tests `./mvnw test`

The error is not reproduced when using Travis (the last build is [ok](https://travis-ci.org/jhipster/generator-jhipster/jobs/380706690)), but it seems the JHipster dependency version is overrided [here](https://github.com/jhipster/generator-jhipster/blob/master/travis/scripts/03-replace-version-generated-project.sh#L15)?

##### **Related issues**

Coming from [this discussion](https://github.com/jhipster/jhipster-automated-samples/pull/5).

##### **Suggest a Fix**

It seems to be linked to the new jhipster-dependencies release: as soon as I change it in `pom.xml`:

- the tests are running with version 2.0.2
- the tests are not running with version 2.0.3

I don't know which upgrade could be responsible.

##### **JHipster Version(s)**

Current version from the master branch

##### **JHipster configuration**

Couchbase as database

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

None.

##### **Browsers and Operating System**

- [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by pascalgrimaud** [comment]
@hdurix : I suppose you already start a Couchbase database ?

In that case, it's normal, as the tests need to start a Couchbase container and it will conflict if there is already one. I mentionned it here: https://github.com/hipster-labs/jhipster-travis-build/issues/2#issuecomment-343761216

That's why in Travis build:
- first, get the Couchbase image: https://github.com/jhipster/generator-jhipster/blob/master/travis/scripts/03-docker-compose.sh#L20-L23
- then launch tests
- before launching the application, start the container: https://github.com/jhipster/generator-jhipster/blob/master/travis/scripts/05-run.sh#L7-L10

So to conclude, it's a real issue, but I didn't found the solution.

**Comment by jdubois** [comment]
I don't know for @hdurix but I did start my container before running my
test :-(

Le ven. 18 mai 2018 à 19:42, Pascal Grimaud <notifications@github.com> a
écrit :

**Comment by pascalgrimaud** [comment]
oh shit, I could reproduce the same issue locally, so strange. My first analyze was wrong

**Comment by pascalgrimaud** [comment]
@hdurix : here the Travis output. The version is not modified

```
$ $JHIPSTER_SCRIPTS/03-replace-version-generated-project.sh
No need to change version in generated project
```

**Comment by hdurix** [comment]
There is definitely something wrong with the port `11210` (maybe with the latest Couchbase version)...

I added this port to the fixed ports list (because for now, the port [is not fixed](https://github.com/differentway/testcontainers-java-module-couchbase/blame/master/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java#L114)) and the tests are now working, see [this commit](https://github.com/hdurix/jhipster-sample-app-couchbase/commit/fd9c7da6622b82bd87c49dd8eabcff4840051cbe) to see the diff.

@pascalgrimaud if you can confirm this trick is working for you as well (I fear that it's only a false positive now), we can ask to add this in TestContainers Couchbase.

@tchlyah any idea?

But it's really annoying I don't understand why this was working before and not anymore (and why it's working with Travis).

BTW it's done the same way in this [blog post](https://dzone.com/articles/testing-spring-data-couchbase-apps-with-testcontainers)

**Comment by jdubois** [comment]
OK same analysis as you @hdurix - and I don't get why it's failing, but there is an issue with this port (which looks correctly configured in testcontainers-couchbase), and fixing it solves the issue. Please note it works on Travis, and I'm using Docker4Mac, so maybe there's an issue with Docker here.

Anyway let's fix the port so it works. I'm also going to upgrade to Couchbase 5.1.0.

@Tcharl I hope it's OK for you!

**Comment by jdubois** [comment]
Done! Thanks @hdurix !!

**Comment by hdurix** [comment]
No problem. It could have been a bit cleaner to do it directly in the lib, so I will do a PR on testcontainers-couchbase and another one for JHipster when the new version of testcontainers-couchbase is released.

**Comment by tchlyah** [comment]
I'm very surprised to see that fixing this port works in travis! I recall I had several issues with this port as being used in travis.

Any way, I didn't have any issues until now, I don't now what have been changed, specially that couchbase-testcontainers tests are working without any issues. Let me analyse this and I'll be back to you.

**Comment by tchlyah** [comment]
Finally it was very quick, I have upgraded [couchbase-testcontainers:1.12](https://github.com/differentway/testcontainers-java-module-couchbase/releases/tag/1.12) and couchbase-client:2.5.7 and now tests build without any issue.

I'll make a PR to jhipster-dependencies. @jdubois can you please upgrade version, and revert your recent commit?

**Comment by jdubois** [comment]
I'm pretty sure we were already using couchbase-testcontainers:1.12 let me check. 
No problem for the release, of course!

**Comment by tchlyah** [comment]
Ah sorry, I made a PR on the wrong library, I didn’t notice that it was deprecated.

After double-checking, yes, the new jhipster-dependency use indeed the new version of testcontainers, but it also uses the new version of [couchbase-client:2.5.8](https://developer.couchbase.com/server/other-products/release-notes-archives/java-sdk#story-h2-2) through springboot dependencies, the later introduce a new fix [JVMCBC-523](https://github.com/couchbase/couchbase-jvm-core/commit/6ed4d47157901b29bddcb586dac882d824a84bf4) that force couchbase-client to retrieve real bucket configuration, and then override the docker mapped port.

I would prefer to avoid as much as possible fixed ports, since it clashes with local couchbase instance as you already experienced, and may cause the tests to fail if this port is already used, as I already experienced with travis.

I will try to find a better way to fix this, and in the mean time, I’ll make a PR, in the right library this time, to fix couchbase-client version to 2.5.7.

If it is OK with you, can you also revert your commit?

**Comment by jdubois** [comment]
@tchlyah I'm thrusting you 100% on that, yes I'm reverting my commit right now

**Comment by jdubois** [comment]
Reverted in https://github.com/jhipster/generator-jhipster/commit/d23779bdd8949ce0bda7afae5c86bd17ab06ea5b

**Comment by jdubois** [comment]
@tchlyah I just released https://github.com/jhipster/jhipster/releases/tag/2.0.4 which includes your couchbase-client upgrade.

This isn't used in jhipster/generator-jhipster yet, as it needs a couple of hours to get synced on Maven Central, so I'll do this later.

**Comment by JulioJu** [comment]
I experience the same problem with jhipster 2.0.6. I test with https://github.com/jhipster/generator-jhipster/blob/master/travis/samples/ngx-couchbase/.yo-rc.json. I don't understand, it should be fixed ?

**Comment by tchlyah** [comment]
It is due to this commit : https://github.com/jhipster/jhipster/commit/db4b2ba0981604b8c490dd627b9245aa40f3bb1d

In my previous comment I explained why we need to stay with the couchbase-client version 2.5.7, until I find a better solution to resolve the issue with new version. I'm working on it.

**Comment by JulioJu** [comment]
@tchlyah maybe you could open again this issue, until it is fix ?

**Comment by tchlyah** [comment]
I've managed to randomize all ports now, even with couchbase-client new version 2.5.8, using some dark magic I'm not very proud of (https://github.com/differentway/testcontainers-java-module-couchbase/commit/5d50da09bffae5a8f35f67c2f3b5c363bfe97aa2). If someone can find a better solution, I'm buying!

I've released [testcontainers-java-module-couchbase:1.2](https://github.com/differentway/testcontainers-java-module-couchbase/releases/tag/1.2), I'll do a PR as soon as the release is available on maven central.
